### PR TITLE
Add indentation rule

### PR DESCRIPTION
**Pull request by jgardella**
Adds a configurable `Indentation` rule to the Typography analyser, which checks that indentation uses a consistent number of spaces (4 by default).

### CODE REVIEW

File: tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs [hunk 1]
```diff
@@ -207,4 +217,24 @@ type TestTypography() =
...
+        this.Parse """
+module P
+
+let x =
+  x"""
+
+        Assert.IsTrue(this.ErrorExistsAt(5, 0))
+
+    [<Test>]
+    member this.``No error for correct indentation``() =
```
Comment (duckmatt) [inline tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs]: Would also be worth adding a test case for places where the lines may not necessarily be indented as per the rest of the file e.g. below on line 3:
```f#
let rainbow =
    { Boss = "Jeffrey"
      Lackeys = ["Zippy"; "George"; "Bungle"] }
```

File: src/FSharpLint.Core/DefaultConfiguration.FSharpLint [hunk 2]
```diff
@@ -327,6 +327,11 @@
           <OneSpaceAllowedAfterOperator>true</OneSpaceAllowedAfterOperator>
           <IgnoreBlankLines>true</IgnoreBlankLines>
         </TrailingWhitespaceOnLine>
+
+        <Indentation>
+          <Enabled>true</Enabled>
```
Comment (duckmatt) [inline src/FSharpLint.Core/DefaultConfiguration.FSharpLint]: I think we should either have this disabled by default or try to auto detect their indentation, otherwise anyone with different valid indentation might have to start adding lint config files (or more likely disable the linter in their editor)

File: tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs (hunk 1 above)
Comment (jgardella) [reply tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs]: Yes this should be added. In fact, my current implementation will fail for this case. I'll add a test for this and try to fix it.

Do you know of anything else which would lead to different indentation like this?

File: src/FSharpLint.Core/DefaultConfiguration.FSharpLint (hunk 2 above)
Comment (jgardella) [reply src/FSharpLint.Core/DefaultConfiguration.FSharpLint]: Agreed.

File: tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs (hunk 1 above)
Comment (jgardella) [reply tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs]: I've updated the implementation and added some tests for this case.

Comment (duckmatt) [reply tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs]: Thanks for the addition of the tests, there's some good docs on indentation here: https://github.com/fsprojects/fantomas/blob/master/docs/FormattingConventions.md#general-rules-for-indentation these could be good test cases. Unfortunately looking at that there's a bunch of additional edge cases :(

Comment (jgardella) [reply tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs]: Thanks for the link Matt, very useful. I looked through it and added handling and test cases for all the exceptional indentation that I saw. Specifically:
- Record field indentation
- Array item indentation
- Pipelining
- Object expressions

There is also the exceptional case of aligning multiple usages of an infix operator, but I'm not sure how to implement this. Did I miss any other cases?

Comment (duckmatt) [reply tests/FSharpLint.Core.Tests/Rules/TestTypographyRules.fs]: No problem. The changes look good, I can't think of any other cases off hand, I think it will be difficult to cover all cases up front, I'll enable it for my projects and see if I spot anything :)
